time_execution/decorator.py

```diff
@@ -1,91 +1,76 @@
-            hooks = settings.hooks + hooks
-
-        # Apply the registered hooks, and collect the metadata they might
-        # return to be stored with the metrics
-        metadata = _apply_hooks(
-            hooks=hooks,
-            response=self.result,
-            exception=self.get_exception(),
-            metric=metric,
-            func=self.func,
-            func_args=self.args,
-            func_kwargs=self.kwargs,
-        )
-
-        metric.update(metadata)
-        write_metric(**metric)
-
-    def get_exception(self):
-        """Retrieve the exception"""
-        if self.exc_info is None:
-            return
-
-        exc_type, exc_value, exc_tb = self.exc_info
-        if exc_value is None:
-            exc_value = exc_type()
-        if exc_value.__traceback__ is not exc_tb:
-            return exc_value.with_traceback(exc_tb)
-        return exc_value
-
-
-class time_execution_async(AsyncDecorator, time_execution):
-    pass
+@overload
+def time_execution(__wrapped: _F) -> _F:
+    """First-order (non-parametrized) decorator with the default FQN getter and hooks by default."""
+
+
+@overload
+def time_execution(
+    *,
+    get_fqn: Callable[[Any], str] = fqn_decorators.get_fqn,
+    extra_hooks: Optional[List] = None,
+    disable_default_hooks: bool = False,
+) -> Callable[[_F], _F]:
+    """
+    Second-order (parametrized) decorator.
+
+    Args:
+        get_fqn: custom FQN getter (uses `fqn-decorators` by default)
+        extra_hooks: additional hooks (next to defined in the settings)
+        disable_default_hooks: if `True`, disable the hooks set by the settings
+    """
+
+
+def time_execution(__wrapped=None, get_fqn: Callable[[Any], str] = fqn_decorators.get_fqn, **kwargs):
+    from time_execution.timed import Timed
+
+    def wrap(__wrapped: _F) -> _F:
+        fqn = get_fqn(__wrapped)
+
+        if not iscoroutinefunction(__wrapped):
+
+            @wraps(__wrapped)
+            def wrapper(*call_args, **call_kwargs):
+                with Timed(wrapped=__wrapped, call_args=call_args, call_kwargs=call_kwargs, fqn=fqn, **kwargs) as timed:
+                    timed.result = __wrapped(*call_args, **call_kwargs)
+                    return timed.result
+
+        else:
+
+            @wraps(__wrapped)
+            async def wrapper(*call_args, **call_kwargs):
+                with Timed(wrapped=__wrapped, call_args=call_args, call_kwargs=call_kwargs, fqn=fqn, **kwargs) as timed:
+                    timed.result = await __wrapped(*call_args, **call_kwargs)
+                    return timed.result
+
+        # Backwards compatibility with `Decorator`.
+        wrapper.fqn = fqn  # type: ignore[attr-defined]
+        wrapper.get_fqn = lambda: wrapper.fqn  # type: ignore[attr-defined]
+        return cast(_F, wrapper)
+
+    return wrap(__wrapped) if __wrapped is not None else wrap
+
+
+# `time_execution` supports async out of the box.
+time_execution_async = time_execution
```

Comment:
Should we deprecate `time_execution_async`?

Reply:
Not sure, I don't mind it here actually